File: apps/api/src/modules/vehicles/domain/vehicle.ts
import type { FuelType, VehicleEventType, VehicleOwnerType } from '@sinaaty/shared-types';
export interface Vehicle {
  id: string; vin: string | null; plateNumber: string | null; plateNumberEn: string | null;
  makeId: number | null; modelId: number | null; makeNameAr: string | null; makeNameEn: string | null; modelNameAr: string | null; modelNameEn: string | null;
  modelYear: number | null; trim: string | null; engine: string | null; fuelType: FuelType | null; colorAr: string | null; odometerKm: number | null;
  ownerType: VehicleOwnerType; ownerUserId: string | null; ownerOrgId: string | null; fleetAssetCode: string | null;
  ownershipVerifiedAt: Date | null; passportPublicToken: string | null; createdAt: Date;
}
export interface VehicleEvent {
  id: string; vehicleId: string; type: VehicleEventType; occurredAt: Date; odometerKm: number | null; orgId: string | null; orgNameAr: string | null;
  refTable: string | null; refId: string | null; summaryAr: string; summaryEn: string | null; data: unknown; isPublic: boolean;
}
export function isOwner(v: Pick<Vehicle, 'ownerType' | 'ownerUserId' | 'ownerOrgId'>, user: { id: string; orgs: Array<{ orgId: string }> }): boolean {
  if (v.ownerType === 'user') return v.ownerUserId === user.id;
  return !!v.ownerOrgId && user.orgs.some((o) => o.orgId === v.ownerOrgId);
}
/** Public passport must never leak owner identity or plate. */
export function toPublicPassport(v: Vehicle, events: VehicleEvent[]) {
  return {
    vehicle: { vin_masked: v.vin ? `${v.vin.slice(0, 8)}*****${v.vin.slice(-4)}` : null, make_ar: v.makeNameAr, make_en: v.makeNameEn, model_ar: v.modelNameAr, model_en: v.modelNameEn, model_year: v.modelYear, fuel_type: v.fuelType, odometer_km: v.odometerKm },
    events: events.filter((e) => e.isPublic).map((e) => ({ type: e.type, occurred_at: e.occurredAt.toISOString(), odometer_km: e.odometerKm, org_name_ar: e.orgNameAr, summary_ar: e.summaryAr, summary_en: e.summaryEn })),
  };
}

File: apps/mobile/lib/features/vehicles/domain/vehicle.dart
class Vehicle {
  final String id; final String? vin; final String? plate; final String? makeAr; final String? modelAr; final int? year; final String? colorAr; final int? odometerKm;
  /// معرّف الصنع — به نُقدّم الورش المتخصّصة بسيارته على غيرها.
  final int? makeId;
  const Vehicle({required this.id, this.vin, this.plate, this.makeAr, this.modelAr, this.year, this.colorAr, this.odometerKm, this.makeId});
  /// «تويوتا كامري 2019» إن عرفنا نوعها، وإلا اللوحة.
  String get name => [makeAr, modelAr, year?.toString()].whereType<String>().join(' ').trim();
  bool get isNamed => name.isNotEmpty;
  String get title => isNamed ? name : (plate ?? vin ?? '');
  /// السطر الثاني لا يكرّر الأول: سيارةٌ أُضيفت بلوحتها وحدها كانت تظهر «أ ب ح 1234» مرّتين —
  /// صفٌّ يقول شيئاً واحداً مرتين ولا يقول ما ينقص. الآن يدعو لإكمال ما يعرّفها.
  String get subtitle {
    if (!isNamed) return vin == null ? 'أضف رقم الهيكل لنعرف نوعها' : 'VIN ${vin!.substring(vin!.length - 6)}';
    return [plate, if (vin != null) 'VIN ${vin!.substring(vin!.length - 6)}'].whereType<String>().join(' · ');
  }
}
class VehicleEvent { final String id; final String type; final DateTime occurredAt; final String summaryAr; final String? summaryEn; final String? orgNameAr; final int? odometerKm; const VehicleEvent({required this.id, required this.type, required this.occurredAt, required this.summaryAr, this.summaryEn, this.orgNameAr, this.odometerKm}); }
class VehiclePassport { final Vehicle vehicle; final List<VehicleEvent> events; const VehiclePassport(this.vehicle, this.events); }
